File: easy/20250630-leetcode-repeated-dna-sequences/main.cpp
// https://leetcode.com/problems/repeated-dna-sequences/

#include <cassert>
#include <unordered_map>
#include <string>
#include <vector>


class Solution {
public:
    std::vector<std::string> findRepeatedDnaSequences(const std::string& s) {
        std::unordered_map<std::string, int> counter; for (int i = 0; i < (int) s.size() - 10 + 1; i++) counter[s.substr(i, 10)] += 1;
        std::vector<std::string> answers; for (const auto& [k, v] : counter) { if (v > 1) answers.push_back(k); }
        return answers;
    }
};


int main() {
    {
        Solution solution = Solution();
        std::vector<std::string> answer = solution.findRepeatedDnaSequences("AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT");
        std::vector<std::string> expected = {"AAAAACCCCC", "CCCCCAAAAA"};
        assert(answer == expected);
    }
    {
        Solution solution = Solution();
        std::vector<std::string> answer = solution.findRepeatedDnaSequences("AAAAAAAAAAAAA");
        std::vector<std::string> expected = {"AAAAAAAAAA"};
        assert(answer == expected);
    }
    {
        Solution solution = Solution();
        std::vector<std::string> answer = solution.findRepeatedDnaSequences("AAAAAAAAAAA");
        std::vector<std::string> expected = {"AAAAAAAAAA"};
        assert(answer == expected);
    }
    return 0;
}

File: easy/20250630-leetcode-repeated-dna-sequences/main.py
# https://leetcode.com/problems/repeated-dna-sequences/

import collections


class Solution:
    def findRepeatedDnaSequences(self, s):
        return [k for k, v in collections.Counter([s[i: i + 10] for i in range(len(s) - 10 + 1)]).items() if v > 1]


solution = Solution()
assert solution.findRepeatedDnaSequences('AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT') == ['AAAAACCCCC', 'CCCCCAAAAA']
assert solution.findRepeatedDnaSequences('AAAAAAAAAAAAA') == ['AAAAAAAAAA']
assert solution.findRepeatedDnaSequences('AAAAAAAAAAA') == ['AAAAAAAAAA']
